# Analysis of Space missions details using sql
####  This project looks into using SQL to Analyze Space data and get insights from the data using sql codes on jupyter notebook 
The data is a csv files of different features collected and i made use of Mysql embedded in Excel tool to stores the data into 
the different tables of the database.
1. Data type: Human space adventure details 
2. Database server type used: **Mysql**

### GOALS
- To create database and tables in the database
- To set constraints such as primary key, foreign key and index
- To changed the data types and also do some data manipulations
- To do some analysis of the data to gain insights into the data


In [1]:
%load_ext sql
#Loading of the sql module so that the environment behaves like sql

In [2]:
%sql mysql://root:praise011020@localhost
#Connecting to the sql server and host

In [3]:
%%sql
/*******************************************************************************
   Space Travel Mission Database --- Version 1.0
   Script: SpaceDetails_Mysql.sql
   Description: Creates and populates the SpaceDetails database.
   DB Server: Mysql
   Author: Babs Fagoroye
   Remodelled: Babajide Fagoroye
   Licence:
********************************************************************************/

/*******************************************************************************
   Drop Foreign Keys Constraints
********************************************************************************/

 * mysql://root:***@localhost
0 rows affected.


[]

In [6]:
%%sql
/******************************************************************************
I am dropping the database and creating a new one so as to avoid redundancy in data
******************************************************************************/

DROP DATABASE IF EXISTS SpaceDetails;

 * mysql://root:***@localhost
3 rows affected.


[]

In [8]:
%%sql
/********************************************************************
I am creating a new database to house the space missions tables and details
********************************************************************/
CREATE DATABASE SpaceDetails;

 * mysql://root:***@localhost
1 rows affected.


[]

In [5]:
%%sql
-- This should gives the output of database present in the DB server.
SHOW DATABASES;

 * mysql://root:***@localhost
8 rows affected.


[('customers_details',),
 ('information_schema',),
 ('mysql',),
 ('performance_schema',),
 ('sakila',),
 ('spacedetails',),
 ('sys',),
 ('world',)]

In [3]:
%%sql
-- Since the Database present in the server is shown then i want to make use of the SpaceDetails which i just created
USE SpaceDetails;

 * mysql://root:***@localhost
0 rows affected.


[]

In [32]:
%%sql
/*****************************************************************************
  Drop Tables              
*****************************************************************************/
/****************************************************************************
 It is always good to drop tables before creating new ones so as not to have 
 multiple duplicate tables
****************************************************************************/
DROP TABLE IF EXISTS CrewedMissions;

DROP TABLE IF EXISTS Missions;

DROP TABLE IF EXISTS SpaceTravelers;

DROP TABLE IF EXISTS SpaceVehicles;

DROP TABLE IF EXISTS Spacewalks;

DROP TABLE IF EXISTS Calender;

 * mysql://root:***@localhost
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.


[]

In [7]:
%%sql
/*******************************************************************************
   Create Tables
********************************************************************************/


CREATE TABLE Calender
(
    Date                    DATE  NOT NULL,
    Year                    VARCHAR(50)  NOT NULL,
    CONSTRAINT PK_CAL       PRIMARY KEY  (Date)
);


CREATE TABLE CrewedMissions
(
    Spacecraft                  VARCHAR(200)  NOT NULL,
    Vehicle                     VARCHAR(160)  NOT NULL,
    Launch                      DATE          NOT NULL,
    Returned                    DATE          NOT NULL,
    Mission                     INTEGER(20)   NOT NULL,
    Status                      CHAR(200)     NOT NULL,
    EarthOrbit                  TEXT(50)      NOT NULL,
    LunarOrbit                  TEXT(50)      NOT NULL,
    Station_visited             CHAR(200),
    Destination                 CHAR(200)     NOT NULL,
    CONSTRAINT PK_CM            PRIMARY KEY  (Spacecraft)        
);

CREATE TABLE Missions (
    Crew_Name           VARCHAR(200) NOT NULL,
    Spacecraft_Launch   VARCHAR(200) NOT NULL,
    Spacecraft_Return   VARCHAR(200) NOT NULL,
    Changed_craft       TEXT(50)     NOT NULL,
    Duration            DECIMAL(10,3) NOT NULL,
    Launch_date         DATE, 
    Return_date         DATE,
    Destination         VARCHAR(200) NOT NULL,
    Space_Station       VARCHAR(200) NOT NULL,
    Mission_Result      VARCHAR(200) NOT NULL
);

CREATE TABLE SpaceTravelers
(
    Name                    VARCHAR(200) NOT NULL,
    Gender                  VARCHAR(50)  NOT NULL,
    Country                 VARCHAR(50)  NOT NULL,
    Continent               VARCHAR(50),
    Birthday                DATE NOT NULL,
    CONSTRAINT PK_ST        PRIMARY KEY  (Name) 
);

CREATE TABLE SpaceVehicles
(
     Rocket                        VARCHAR(200) NOT NULL,
     Active                        CHAR(150)  NOT NULL,
     Country                       CHAR(150)  NOT NULL,
     LEO_payload                   INTEGER(50),
     GTO_payload                   INTEGER(50),
     TLI_Mars_payload              INTEGER(50),
     Liftoff_Height                DECIMAL(7,3),
     Liftoff_Mass                  INTEGER(50),
     Successful_flights            INTEGER(50),
     Total_flights                 INTEGER(50),
     FirstLaunch                   DATE,
     LastLaunch                    DATE,
     Years                         VARCHAR(50),
     Class                         VARCHAR(150),
     Carried_crews                 BOOLEAN
);

CREATE TABLE Spacewalks
(
    CrewMissionName                    VARCHAR(200) NOT NULL,
    Name                               VARCHAR(200) NOT NULL,
    EVA_Number                         INTEGER(50)  NOT NULL,
    StartDate                          DATE,
    EndDate                            DATE,
    Duration                           TIME,
    Moonwalk                           BOOLEAN,
    StandUpOnly                        BOOLEAN,
    DeepSpace                          BOOLEAN,
    Traveler                           DATE,
    Launching_spacecraft               VARCHAR(150),
    Destination                        VARCHAR(150)
);



 * mysql://root:***@localhost
0 rows affected.


[]

In [ ]:
%%sql
/******************************************************************
I am creating a virtual table from the preexisting tables.
*****************************************************************/

CREATE TABLE NextedTable AS
SELECT Destination,  COUNT(*) as count
FROM (SELECT ROW_NUMBER() OVER(ORDER BY Crew_Name) AS Indexing, Destination
FROM 
Missions) Nexted
GROUP BY 1
ORDER BY 2 DESC ;


In [8]:
%%sql
-- To know the tables present in the SpaceDeatiails DB created.
SHOW TABLES

 * mysql://root:***@localhost
7 rows affected.


[('calender',),
 ('crewedmissions',),
 ('missions',),
 ('nextedtable',),
 ('spacetravelers',),
 ('spacevehicles',),
 ('spacewalks',)]

#### The values in each tables is inserted from Microsoft Excel by embedding mysql extention into the software.
I did it so that the values would be automatically refresh or change as i update the values from excel 

In [4]:
%%sql
DESCRIBE Calender;

 * mysql://root:***@localhost
2 rows affected.


[('Date', 'date', 'NO', 'PRI', None, ''),
 ('Year', 'varchar(50)', 'NO', 'MUL', None, '')]

In [5]:
%%sql
-- Show the result of the newly created table. To see if it gives the desired results.

SELECT * FROM
NextedTable

 * mysql://root:***@localhost
6 rows affected.


[('Earth Orbit', 597),
 ('ISS', 422),
 ('Mir', 137),
 ('Other Space station', 90),
 ('Moon', 27),
 ('Non-orbital space', 18)]

In [6]:
%%sql
-- To view the data in the table CrewedMissions and be sure the table worked. I set a limit of 5 so as to show 5 rows only 

SELECT *
FROM CrewedMissions
LIMIT 5;

 * mysql://root:***@localhost
5 rows affected.


[('41-B, Challenger', 'Space Shuttle', datetime.date(1984, 2, 3), datetime.date(1984, 2, 11), 98, 'Returned', 'True', 'False', None, 'Earth Orbit'),
 ('41-C, Challenger', 'Space Shuttle', datetime.date(1984, 4, 6), datetime.date(1984, 4, 13), 101, 'Returned', 'True', 'False', None, 'Earth Orbit'),
 ('41-D, Discovery', 'Space Shuttle', datetime.date(1984, 8, 30), datetime.date(1984, 9, 5), 103, 'Returned', 'True', 'False', None, 'Earth Orbit'),
 ('41-G, Challenger', 'Space Shuttle', datetime.date(1984, 10, 5), datetime.date(1984, 10, 13), 104, 'Returned', 'True', 'False', None, 'Earth Orbit'),
 ('51-A, Discovery', 'Space Shuttle', datetime.date(1984, 11, 8), datetime.date(1984, 11, 16), 105, 'Returned', 'True', 'False', None, 'Earth Orbit')]

In [7]:
%%sql
-- To also see the values in the table SpaceVehicles

SELECT * 
FROM Spacevehicles
LIMIT 5;

 * mysql://root:***@localhost
5 rows affected.


[('Sputnik\xa08K71PS', 'RETIRED', 'Russia', 500, None, None, Decimal('30.000'), 267000, 2, 2, datetime.date(1957, 10, 4), datetime.date(1957, 11, 3), '1957', 'SLV', 0),
 ('Vanguard', 'RETIRED', 'United States', 11, None, None, Decimal('21.900'), 10050, 3, 11, datetime.date(1957, 12, 6), datetime.date(1959, 9, 18), '1957', 'SLV', 0),
 ('Sputnik\xa08A91', 'RETIRED', 'Russia', 1327, None, None, Decimal('31.100'), 269300, 1, 2, datetime.date(1958, 4, 27), datetime.date(1958, 5, 15), '1958', 'SLV', 0),
 ('NOTS-EV-1 Pilot', 'RETIRED', 'United States', 1, None, None, Decimal('4.400'), 950, 0, 10, datetime.date(1958, 7, 4), datetime.date(1958, 8, 17), '1958', 'SLV', 0),
 ('Scout', 'RETIRED', 'United States', 174, None, None, Decimal('25.000'), 17850, 87, 99, datetime.date(1960, 4, 18), datetime.date(1994, 5, 9), '1960', 'SLV', 0)]

In [8]:
%%sql
-- Doing same as of the above too for the calendar table.

SELECT * FROM
Calender 
LIMIT 5;

 * mysql://root:***@localhost
5 rows affected.


[(datetime.date(1960, 1, 1), '1960'),
 (datetime.date(1960, 1, 2), '1960'),
 (datetime.date(1960, 1, 3), '1960'),
 (datetime.date(1960, 1, 4), '1960'),
 (datetime.date(1960, 1, 5), '1960')]

In [ ]:
%%sql
/**************************************************************************
creating a foreign keys on tables that have columns with similar attributes  
**************************************************************************/

ALTER TABLE Missions
ADD CONSTRAINT FK_MDST
FOREIGN KEY (Crew_Name)  REFERENCES SpaceTravelers(Name);

ALTER TABLE Spacewalks
ADD CONSTRAINT FK_SWST
FOREIGN KEY (Name)  REFERENCES SpaceTravelers(Name);

ALTER TABLE Missions
ADD CONSTRAINT FK_MNXT
FOREIGN KEY (Destination)  REFERENCES NextedTable(Destination);

ALTER TABLE CrewedMissions
ADD CONSTRAINT FK_CMNXT
FOREIGN KEY (Destination)  REFERENCES NextedTable(Destination);

ALTER TABLE Spacewalks
ADD CONSTRAINT FK_SWNXT
FOREIGN KEY (Destination)  REFERENCES NextedTable(Destination);

In [ ]:
%%sql 
/**************************************************************************
creating a foreign keys on tables that have columns with similar attributes  
**************************************************************************/

ALTER TABLE Missions 
ADD CONSTRAINT FK_MS
FOREIGN KEY (Launch_date)  REFERENCES Calender(Date)
ON DELETE NO ACTION ON UPDATE NO ACTION;


ALTER TABLE Spacewalks
ADD CONSTRAINT FK_SPW
FOREIGN KEY (StartDate)  REFERENCES Calender(Date);

ALTER TABLE CrewedMissions 
ADD CONSTRAINT FK_CM
FOREIGN KEY (Launch)  REFERENCES Calender(Date)
ON DELETE NO ACTION ON UPDATE NO ACTION;


In [9]:
#%%sql
#SELECT
#DATE_FORMAT(STR_TO_DATE(StartDate,'%m/%d/%Y %H:%m:%s'),'%Y-%m-%d %H:%i:%s') HAIL
#FROM
#Spacewalks

In [37]:
#%%sql
#UPDATE Calender
#SET Date = DATE_FORMAT(STR_TO_DATE(Date,'%m/%d/%Y %H'),'%Y-%m-%d %H:%i:%s');

In [10]:
%%sql
DESCRIBE Spacevehicles


 * mysql://root:***@localhost
15 rows affected.


[('Rocket', 'varchar(200)', 'NO', '', None, ''),
 ('Active', 'char(150)', 'NO', '', None, ''),
 ('Country', 'char(150)', 'NO', '', None, ''),
 ('LEO_payload', 'int', 'YES', '', None, ''),
 ('GTO_payload', 'int', 'YES', '', None, ''),
 ('TLI_Mars_payload', 'int', 'YES', '', None, ''),
 ('Liftoff_Height', 'decimal(7,3)', 'YES', '', None, ''),
 ('Liftoff_Mass', 'int', 'YES', '', None, ''),
 ('Successful_flights', 'int', 'YES', '', None, ''),
 ('Total_flights', 'int', 'YES', '', None, ''),
 ('FirstLaunch', 'date', 'YES', '', None, ''),
 ('LastLaunch', 'date', 'YES', '', None, ''),
 ('Years', 'varchar(50)', 'YES', '', None, ''),
 ('Class', 'varchar(150)', 'YES', '', None, ''),
 ('Carried_crews', 'tinyint(1)', 'YES', '', None, '')]

In [10]:
%%sql
SELECT * 
FROM 
SpaceVehicles
LIMIT 5;

 * mysql://root:***@localhost
5 rows affected.


[('Angara 1.2', 'ACTIVE', 'Russia', 3800, 1430, None, Decimal('42.700'), 171500, 1, 1, datetime.date(2014, 7, 9), datetime.date(2014, 7, 9), 'MLV', 0),
 ('Angara A5', 'ACTIVE', 'Russia', 24500, 7500, None, Decimal('64.000'), 790000, 1, 1, datetime.date(2014, 7, 23), datetime.date(2014, 7, 23), 'HLV', 0),
 ('Antares 110', 'RETIRED', 'United States', 8000, 1500, None, Decimal('40.500'), 282000, 2, 2, datetime.date(2013, 4, 21), datetime.date(2013, 9, 18), 'MLV', 0),
 ('Antares 120', 'RETIRED', 'United States', 8000, 1500, None, Decimal('40.500'), 282000, 2, 2, datetime.date(2014, 1, 9), datetime.date(2014, 7, 13), 'MLV', 0),
 ('Antares 130', 'RETIRED', 'United States', 8000, 1500, None, Decimal('41.900'), 296000, 0, 1, datetime.date(2014, 10, 28), datetime.date(2014, 10, 28), 'MLV', 0)]

In [13]:
%%sql
-- This is to change the format of the Date in the table Spacewalks to suit the needed format

SELECT
DATE_FORMAT(STR_TO_DATE(StartDate, '%m/%d/%Y %H'), '%m-%d-%Y %H:%i:%s') AS Start
FROM
Spacewalks;

In [14]:
%%sql
-- Show 10 rows of the table Spacewalks so has to see the results

SELECT * 
FROM Spacewalks
LIMIT 10;

 * mysql://root:***@localhost
10 rows affected.


[('Voskhod 2', 'Alexei Leonov', 1, datetime.date(1965, 3, 18), datetime.date(1965, 3, 18), datetime.timedelta(seconds=729), 0, 0, 0, datetime.date(1965, 3, 18), 'Voskhod 2', 'Earth Orbit'),
 ('Gemini 4', 'Ed White', 2, datetime.date(1965, 6, 3), datetime.date(1965, 6, 3), datetime.timedelta(seconds=1200), 0, 0, 0, datetime.date(1965, 6, 3), 'Gemini 4', 'Earth Orbit'),
 ('Gemini 9', 'Gene Cernan', 3, datetime.date(1966, 6, 5), datetime.date(1966, 6, 5), datetime.timedelta(seconds=7620), 0, 0, 0, datetime.date(1966, 6, 3), 'Gemini 9', 'Earth Orbit'),
 ('Gemini 10', 'Michael Collins', 4, datetime.date(1966, 7, 19), datetime.date(1966, 7, 19), datetime.timedelta(seconds=2940), 0, 0, 0, datetime.date(1966, 7, 18), 'Gemini 10', 'Earth Orbit'),
 ('Gemini 10', 'Michael Collins', 5, datetime.date(1966, 7, 20), datetime.date(1966, 7, 20), datetime.timedelta(seconds=2340), 0, 0, 0, datetime.date(1966, 7, 18), 'Gemini 10', 'Earth Orbit'),
 ('Gemini 11', 'Richard Gordon', 6, datetime.date(1966, 9, 13), datetime.date(1966, 9, 13), datetime.timedelta(seconds=1980), 0, 0, 0, datetime.date(1966, 9, 12), 'Gemini 11', 'Earth Orbit'),
 ('Gemini 11', 'Richard Gordon', 7, datetime.date(1966, 9, 14), datetime.date(1966, 9, 14), datetime.timedelta(seconds=7680), 0, 0, 0, datetime.date(1966, 9, 12), 'Gemini 11', 'Earth Orbit'),
 ('Gemini 12', 'Buzz Aldrin', 8, datetime.date(1966, 11, 12), datetime.date(1966, 11, 12), datetime.timedelta(seconds=8940), 0, 0, 0, datetime.date(1966, 11, 11), 'Gemini 12', 'Earth Orbit'),
 ('Gemini 12', 'Buzz Aldrin', 9, datetime.date(1966, 11, 13), datetime.date(1966, 11, 13), datetime.timedelta(seconds=7560), 0, 0, 0, datetime.date(1966, 11, 11), 'Gemini 12', 'Earth Orbit'),
 ('Gemini 12', 'Buzz Aldrin', 10, datetime.date(1966, 11, 14), datetime.date(1966, 11, 14), datetime.timedelta(seconds=3300), 0, 0, 0, datetime.date(1966, 11, 11), 'Gemini 12', 'Earth Orbit')]

In [15]:
#%%sql
#ALTER TABLE SpaceVehicles 
#ADD CONSTRAINT FK_SV
#FOREIGN KEY (Last_launch)  REFERENCES Calender(Date)     -- Just for reference purpose only
#ON DELETE NO ACTION ON UPDATE NO ACTION
#ON DELETE CASCADE ON UPDATE CASCADE
#ON DELETE SET NULL ON UPDATE SET NULL;

#SET foreign_key_checks = 0
#SET foreign_key_checks = 1;

In [15]:
%%sql
/*****************************************
Activating the foreign keys to work. To turn off set global would be equal to zero(0) instead of one(1)
*****************************************/

SET GLOBAL FOREIGN_KEY_CHECKS=1;

 * mysql://root:***@localhost
0 rows affected.


[]

In [16]:
%%sql
/***************************************************
To see the constraints attached to tables based on a
particular referenced table(calendar table)now.
****************************************************/

SELECT TABLE_NAME, COLUMN_NAME, CONSTRAINT_NAME,
REFERENCED_TABLE_NAME, REFERENCED_COLUMN_NAME
FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
WHERE REFERENCED_TABLE_NAME = 'Calender';


 * mysql://root:***@localhost
3 rows affected.


[('missions', 'Launch_date', 'FK_MS', 'calender', 'Date'),
 ('crewedmissions', 'Launch', 'FK_CM', 'calender', 'Date'),
 ('spacewalks', 'StartDate', 'FK_SPW', 'calender', 'Date')]

In [17]:
%%sql
/***************************************************
To see the constraints attached to tables based on a
particular referenced table(SpaceTravelers table)now.
****************************************************/

SELECT TABLE_NAME, COLUMN_NAME, CONSTRAINT_NAME,
REFERENCED_TABLE_NAME, REFERENCED_COLUMN_NAME
FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
WHERE REFERENCED_TABLE_NAME = 'SpaceTravelers';


 * mysql://root:***@localhost
2 rows affected.


[('missions', 'Crew_Name', 'FK_MDST', 'spacetravelers', 'Name'),
 ('spacewalks', 'Name', 'FK_SWST', 'spacetravelers', 'Name')]

In [18]:
%%sql
/***************************************************
To see the constraints attached to tables based on a
particular referenced table(NextedTable table)now.
****************************************************/

SELECT TABLE_NAME, COLUMN_NAME, CONSTRAINT_NAME,
REFERENCED_TABLE_NAME, REFERENCED_COLUMN_NAME
FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
WHERE REFERENCED_TABLE_NAME = 'NextedTable';


 * mysql://root:***@localhost
3 rows affected.


[('missions', 'Destination', 'FK_MNXT', 'nextedtable', 'Destination'),
 ('crewedmissions', 'Destination', 'FK_CMNXT', 'nextedtable', 'Destination'),
 ('spacewalks', 'Destination', 'FK_SWNXT', 'nextedtable', 'Destination')]

In [21]:
%%sql
/****************To check the character set of the data in tables aforementioned****************/

SELECT CCSA.character_set_name
FROM information_schema.TABLES T,
information_schema.COLLATION_CHARACTER_SET_APPLICABILITY CCSA
WHERE CCSA.collation_name = T.table_collation 
AND T.table_schema = 'SpaceDetails'
AND T.table_name = 'SpaceVehicles'

 * mysql://root:***@localhost
1 rows affected.


[('utf8mb4',)]

In [20]:
%%sql
/*********TO change character set of Database and Table*************/

ALTER DATABASE SpaceDetails
CHARACTER SET utf8mb4 
COLLATE utf8mb4_unicode_ci;


ALTER TABLE SpaceVehicles
CONVERT TO CHARACTER 
SET utf8mb4 
COLLATE utf8mb4_unicode_ci;

 * mysql://root:***@localhost
1 rows affected.


[]

In [23]:
%%sql
SHOW TABLES;

 * mysql://root:***@localhost
7 rows affected.


[('calender',),
 ('crewedmissions',),
 ('missions',),
 ('nextedtable',),
 ('spacetravelers',),
 ('spacevehicles',),
 ('spacewalks',)]

In [29]:
%%sql
/****************************************************************
Creating index constraints on the calender and NextedTable tables
so as to allow easy paring of foreign keys to other tables
***************************************************************/

CREATE INDEX IFK_Cal ON Calender(Year);

CREATE INDEX IFK_NXT ON NextedTable(Destination);

In [17]:
%%sql
-- To see how the table missions looks before running some aggregations on it.
SELECT * 
FROM 
Missions
LIMIT 5;

 * mysql://root:***@localhost
5 rows affected.


[('Yuri Gagarin', 'Vostok 1', 'Vostok 1', 'False', Decimal('0.070'), datetime.date(1961, 4, 21), datetime.date(1961, 4, 21), 'Earth Orbit', 'None visited', 'Returned'),
 ('Alan Shepard', 'Mercury-Redstone 3', 'Mercury-Redstone 3', 'False', Decimal('0.010'), datetime.date(1961, 5, 5), datetime.date(1961, 5, 5), 'Non-orbital space', 'None visited', 'Returned'),
 ('Virgil Grissom', 'Mercury-Redstone 4', 'Mercury-Redstone 4', 'False', Decimal('0.010'), datetime.date(1961, 7, 21), datetime.date(1961, 7, 21), 'Non-orbital space', 'None visited', 'Returned'),
 ('Gherman Titov', 'Vostok 2', 'Vostok 2', 'False', Decimal('1.050'), datetime.date(1961, 8, 6), datetime.date(1961, 8, 7), 'Earth Orbit', 'None visited', 'Returned'),
 ('John Glenn', 'Mercury-Atlas 6', 'Mercury-Atlas 6', 'False', Decimal('0.210'), datetime.date(1962, 2, 20), datetime.date(1962, 2, 20), 'Earth Orbit', 'None visited', 'Returned')]

#### I want to get the following informatiion from the data in the tables
1. The Crew members names and the location most visited (setting bench mark of visitation to 500 or more times)
2. Countries with the highest male space Travelers
3. Countries with the highest female space Travelers
4. Total amount of flights by class.
5. Active Rockets produced in Russia that have loaded 3000 and about loads.
6. Crew members name and the duration difference between their first and last flights.
7. Details of Crew members whose EVA number is between 3 and 15.

In [35]:
%%sql
SELECT M.Crew_Name, NT.Destination Location
FROM NextedTable NT
JOIN Missions M
ON
NT.Destination  =  M.Destination
WHERE Count >= 500 

 * mysql://root:***@localhost
597 rows affected.


[('Yuri Gagarin', 'Earth Orbit'),
 ('Gherman Titov', 'Earth Orbit'),
 ('John Glenn', 'Earth Orbit'),
 ('Scott Carpenter', 'Earth Orbit'),
 ('Andriyan Nikolayev', 'Earth Orbit'),
 ('Pavel Popovich', 'Earth Orbit'),
 ('Wally Schirra', 'Earth Orbit'),
 ('Gordon Cooper', 'Earth Orbit'),
 ('Valery Bykovsky', 'Earth Orbit'),
 ('Valentina Tereshkova', 'Earth Orbit'),
 ('Boris Yegorov', 'Earth Orbit'),
 ('Konstantin Feoktistov', 'Earth Orbit'),
 ('Vladimir Komarov', 'Earth Orbit'),
 ('Alexei Leonov', 'Earth Orbit'),
 ('Pavel Belyayev', 'Earth Orbit'),
 ('John Young', 'Earth Orbit'),
 ('Virgil Grissom', 'Earth Orbit'),
 ('Ed White', 'Earth Orbit'),
 ('James McDivitt', 'Earth Orbit'),
 ('Gordon Cooper', 'Earth Orbit'),
 ('Pete Conrad', 'Earth Orbit'),
 ('Frank Borman', 'Earth Orbit'),
 ('Jim Lovell', 'Earth Orbit'),
 ('Thomas P. Stafford', 'Earth Orbit'),
 ('Wally Schirra', 'Earth Orbit'),
 ('David Scott', 'Earth Orbit'),
 ('Neil Armstrong', 'Earth Orbit'),
 ('Gene Cernan', 'Earth Orbit'),
 ('Thomas P. Stafford', 'Earth Orbit'),
 ('John Young', 'Earth Orbit'),
 ('Michael Collins', 'Earth Orbit'),
 ('Pete Conrad', 'Earth Orbit'),
 ('Richard Gordon', 'Earth Orbit'),
 ('Buzz Aldrin', 'Earth Orbit'),
 ('Jim Lovell', 'Earth Orbit'),
 ('Vladimir Komarov', 'Earth Orbit'),
 ('Donn F. Eisele', 'Earth Orbit'),
 ('Wally Schirra', 'Earth Orbit'),
 ('Walter Cunningham', 'Earth Orbit'),
 ('Georgi Beregovoi', 'Earth Orbit'),
 ('Vladimir Shatalov', 'Earth Orbit'),
 ('Aleksei Yeliseyev', 'Earth Orbit'),
 ('Boris Volynov', 'Earth Orbit'),
 ('Yevgeny Khrunov', 'Earth Orbit'),
 ('David Scott', 'Earth Orbit'),
 ('James McDivitt', 'Earth Orbit'),
 ('Rusty Schweickart', 'Earth Orbit'),
 ('Georgy Shonin', 'Earth Orbit'),
 ('Valery Kubasov', 'Earth Orbit'),
 ('Anatoly Filipchenko', 'Earth Orbit'),
 ('Viktor Gorbatko', 'Earth Orbit'),
 ('Vladislav Volkov', 'Earth Orbit'),
 ('Aleksei Yeliseyev', 'Earth Orbit'),
 ('Vladimir Shatalov', 'Earth Orbit'),
 ('Andriyan Nikolayev', 'Earth Orbit'),
 ('Vitaly Sevastyanov', 'Earth Orbit'),
 ('Oleg Makarov', 'Earth Orbit'),
 ('Vasili Lazarev', 'Earth Orbit'),
 ('Pyotr Klimuk', 'Earth Orbit'),
 ('Valentin Lebedev', 'Earth Orbit'),
 ('Gennadi Sarafanov', 'Earth Orbit'),
 ('Lev Dyomin', 'Earth Orbit'),
 ('Anatoly Filipchenko', 'Earth Orbit'),
 ('Nikolay Rukavishnikov', 'Earth Orbit'),
 ('Alexei Leonov', 'Earth Orbit'),
 ('Valery Kubasov', 'Earth Orbit'),
 ('Donald Slayton', 'Earth Orbit'),
 ('Thomas P. Stafford', 'Earth Orbit'),
 ('Vance Brand', 'Earth Orbit'),
 ('Valery Bykovsky', 'Earth Orbit'),
 ('Vladimir Aksyonov', 'Earth Orbit'),
 ('Valery Rozhdestvensky', 'Earth Orbit'),
 ('Vyacheslav Zudov', 'Earth Orbit'),
 ('Valery Ryumin', 'Earth Orbit'),
 ('Vladimir Kovalyonok', 'Earth Orbit'),
 ('Georgi Ivanov', 'Earth Orbit'),
 ('Nikolay Rukavishnikov', 'Earth Orbit'),
 ('John Young', 'Earth Orbit'),
 ('Robert Crippen', 'Earth Orbit'),
 ('Joseph H. Engle', 'Earth Orbit'),
 ('Richard H. Truly', 'Earth Orbit'),
 ('C. Gordon Fullerton', 'Earth Orbit'),
 ('Jack Lousma', 'Earth Orbit'),
 ('Henry Hartsfield', 'Earth Orbit'),
 ('Thomas Kenneth Mattingly II', 'Earth Orbit'),
 ('Joseph P. Allen', 'Earth Orbit'),
 ('Robert Overmyer', 'Earth Orbit'),
 ('Vance Brand', 'Earth Orbit'),
 ('William B. Lenoir', 'Earth Orbit'),
 ('Donald Peterson', 'Earth Orbit'),
 ('Karol Bobko', 'Earth Orbit'),
 ('Paul Weitz', 'Earth Orbit'),
 ('Story Musgrave', 'Earth Orbit'),
 ('Aleksandr Serebrov', 'Earth Orbit'),
 ('Gennadi Strekalov', 'Earth Orbit'),
 ('Vladimir Titov', 'Earth Orbit'),
 ('Frederick Hauck', 'Earth Orbit'),
 ('John Fabian', 'Earth Orbit'),
 ('Norman Thagard', 'Earth Orbit'),
 ('Robert Crippen', 'Earth Orbit'),
 ('Sally Ride', 'Earth Orbit'),
 ('Dale Gardner', 'Earth Orbit'),
 ('Daniel Brandenstein', 'Earth Orbit'),
 ('Guion Bluford', 'Earth Orbit'),
 ('Richard H. Truly', 'Earth Orbit'),
 ('William Thornton', 'Earth Orbit'),
 ('Brewster Shaw', 'Earth Orbit'),
 ('Byron Lichtenberg', 'Earth Orbit'),
 ('John Young', 'Earth Orbit'),
 ('Owen Garriott

###### From the above, 597 crew members travelled on missions 500 times or more and their main destinations is Earth Orbit

In [37]:
%%sql

SELECT P.Country, COUNT(P.Country) NO_MALES
FROM
(SELECT M.Crew_Name, ST.Country, COUNT(ST.Gender) GENDER
FROM
SpaceTravelers ST
LEFT JOIN Missions M
ON M.Crew_Name = ST.Name
WHERE Gender =  "Male"
GROUP BY 1
ORDER BY 3 DESC) P
GROUP BY 1
ORDER BY 2 DESC;

 * mysql://root:***@localhost
41 rows affected.


[('United States', 291),
 ('Russia', 118),
 ('Germany', 11),
 ('Japan', 10),
 ('France', 9),
 ('China', 9),
 ('Canada', 7),
 ('Italy', 6),
 ('United Kingdom', 3),
 ('Netherlands', 3),
 ('Bulgaria', 2),
 ('Australia', 2),
 ('Belgium', 2),
 ('Spain', 2),
 ('Kazakhstan', 2),
 ('Hungary', 2),
 ('Switzerland', 1),
 ('Sweden', 1),
 ('Armenia', 1),
 ('Afghanistan', 1),
 ('Costa Rica', 1),
 ('Denmark', 1),
 ('Cuba', 1),
 ('Iceland', 1),
 ('Romania', 1),
 ('Austria', 1),
 ('United Arab Emirates', 1),
 ('Israel', 1),
 ('Slovakia', 1),
 ('Mongolia', 1),
 ('Ukraine', 1),
 ('Brazil', 1),
 ('South Africa', 1),
 ('Poland', 1),
 ('Syria', 1),
 ('Vietnam', 1),
 ('India', 1),
 ('Mexico', 1),
 ('Malaysia', 1),
 ('Saudi Arabia', 1),
 ('Czech Republic', 1)]

##### From the data result above, it shows that `USA` produced more male space travelers than the rest of the countries.
with `Russia, Germany, Japan, and France` making the top 5.

In [39]:
%%sql

SELECT M.Crew_Name, ST.Country, COUNT(ST.Gender) GENDER
FROM
SpaceTravelers ST
LEFT JOIN Missions M
ON M.Crew_Name = ST.Name
WHERE Gender =  "Female"
GROUP BY 1
ORDER BY 3 DESC

 * mysql://root:***@localhost
65 rows affected.


[('Bonnie J. Dunbar', 'United States', 5),
 ('Tamara E. Jernigan', 'United States', 5),
 ('Susan Helms', 'United States', 5),
 ('Shannon Lucid', 'United States', 5),
 ('Janice E. Voss', 'United States', 5),
 ('Marsha Ivins', 'United States', 5),
 ('Linda Godwin', 'United States', 4),
 ('Wendy Lawrence', 'United States', 4),
 ('Nancy J. Currie-Gregg', 'United States', 4),
 ('Kathryn C. Thornton', 'United States', 4),
 ('Eileen Collins', 'United States', 4),
 ('Ellen Ochoa', 'United States', 4),
 ('Janet L. Kavandi', 'United States', 3),
 ('Pamela Melroy', 'United States', 3),
 ('Catherine Coleman', 'United States', 3),
 ('Nancy Jan Davis', 'United States', 3),
 ('Ellen Baker', 'United States', 3),
 ('Kathryn D. Sullivan', 'United States', 3),
 ('Stephanie Wilson', 'United States', 3),
 ('Sandra Magnus', 'United States', 3),
 ('Peggy Whitson', 'United States', 3),
 ('Margaret Rhea Seddon', 'United States', 3),
 ('Tracy Caldwell Dyson', 'United States', 2),
 ('Yelena Kondakova', 'Russia', 2),
 ('Chiaki Mukai', 'Japan', 2),
 ('Karen Nyberg', 'United States', 2),
 ('Claudie Haigneré', 'France', 2),
 ('Heidemarie Stefanyshyn-Piper', 'United States', 2),
 ('Judith Resnik', 'United States', 2),
 ('Julie Payette', 'Canada', 2),
 ('Kalpana Chawla', 'India', 2),
 ('Kathryn Hire', 'United States', 2),
 ('Svetlana Savitskaya', 'Russia', 2),
 ('Mary Cleave', 'United States', 2),
 ('Mary Weber', 'United States', 2),
 ('Nicole Stott', 'United States', 2),
 ('Sally Ride', 'United States', 2),
 ('Sunita Williams', 'United States', 2),
 ('Susan Kilrain', 'United States', 2),
 ('Anne McClain', 'United States', 1),
 ('Anousheh Ansari', 'Iran', 1),
 ('Naoko Yamazaki', 'Japan', 1),
 ('Laurel B. Clark', 'United States', 1),
 ('K. Megan McArthur', 'United States', 1),
 ('Joan Higginbotham', 'United States', 1),
 ('Roberta Bondar', 'Canada', 1),
 ('Mae Jemison', 'United States', 1),
 ('Samantha Cristoforetti', 'Italy', 1),
 ('Jessica Meir', 'United States', 1),
 ('Serena Auñón-Chancellor', 'United States', 1),
 ('Christina Koch', 'United States', 1),
 ('Shannon Walker', 'United States', 1),
 ('Helen Sharman', 'United Kingdom', 1),
 ('Lisa Nowak', 'United States', 1),
 ('Dorothy Metcalf-Lindenburger', 'United States', 1),
 ('Millie Hughes-Fulford', 'United States', 1),
 ('Anna Fisher', 'United States', 1),
 ('Christa McAuliffe', 'United States', 1),
 ('Kathleen Rubins', 'United States', 1),
 ('Valentina Tereshkova', 'Russia', 1),
 ('Wang Yaping', 'China', 1),
 ('Barbara Morgan', 'United States', 1),
 ('Liu Yang', 'China', 1),
 ('Yelena Serova', 'Russia', 1),
 ('Yi So-yeon', 'South Korea', 1)]

##### From the result it shows that `USA` has more representation in the female categories also followed by `Russia`

In [51]:
%%sql

SELECT Class,Sum(Total_Flights) Flights
FROM 
SpaceVehicles
GROUP BY Class
ORDER BY Flights DESC;

 * mysql://root:***@localhost
5 rows affected.


[('MLV', Decimal('3812')),
 ('SLV', Decimal('993')),
 ('HLV', Decimal('483')),
 ('---', Decimal('156')),
 ('SHLV', Decimal('22'))]

##### The result above shows that Class *`MLV`* have the highest flights with 3812 flights taken via that class

In [44]:
%%sql

SELECT Rocket, Country, COUNT(DISTINCT(Active)) SPACTIVE, LEO_Payload 
FROM 
SpaceVehicles
WHERE Country = "Russia"
GROUP BY 1
HAVING SUM(LEO_Payload)>=3000
ORDER BY 4 DESC
LIMIT 20
;

 * mysql://root:***@localhost
20 rows affected.


[('N1', 'Russia', 1, 95000),
 ('Energia', 'Russia', 1, 88000),
 ('Angara A5', 'Russia', 1, 24500),
 ('Proton-M', 'Russia', 1, 23000),
 ('Proton-K', 'Russia', 1, 19760),
 ('Soyuz-2.1b', 'Russia', 1, 8200),
 ('Soyuz-U2', 'Russia', 1, 7050),
 ('Soyuz ST-A', 'Russia', 1, 7020),
 ('Soyuz ST-B', 'Russia', 1, 7020),
 ('Soyuz-2.1a', 'Russia', 1, 7020),
 ('Soyuz-FG', 'Russia', 1, 6900),
 ('Soyuz-U', 'Russia', 1, 6900),
 ('Soyuz-M', 'Russia', 1, 6600),
 ('Soyuz', 'Russia', 1, 6450),
 ('Voskhod', 'Russia', 1, 6000),
 ('Soyuz-L', 'Russia', 1, 5500),
 ('Vostok-2', 'Russia', 1, 4730),
 ('Soyuz/Vostok', 'Russia', 1, 4500),
 ('Vostok-L', 'Russia', 1, 4500),
 ('Angara 1.2', 'Russia', 1, 3800)]

##### It shows that Rocket named N1 has the highest LEO payload of 95000

In [46]:
%%sql
-- Here i created a CTE so as to harmonized the tables and to ease my findings
-- I also created a lead in duration and also replace any null value with 0 or NV

WITH CrewDetails AS 
(SELECT CM.Launch,M.Crew_Name, CM.Destination,CM.Station_Visited,M.Duration
FROM CrewedMissions CM
LEFT JOIN Missions M
ON CM.Launch = M.Launch_Date
AND CM.Destination = M.Destination)

SELECT *, COALESCE((Duration - LEAD(Duration,1) 
 OVER (PARTITION BY Crew_Name ORDER BY Launch)),0) AS Diff,  COALESCE(Station_Visited, "NV") AS StationV
FROM 
CrewDetails
LIMIT 100;


 * mysql://root:***@localhost
100 rows affected.


[(datetime.date(1979, 6, 6), None, 'Other Space station', 'Salyut 6', None, Decimal('0.000'), 'Salyut 6'),
 (datetime.date(1988, 8, 29), 'Abdul Ahad Mohmand', 'Mir', 'Mir', Decimal('9.170'), Decimal('0.000'), 'Mir'),
 (datetime.date(2008, 5, 31), 'Akihiko Hoshide', 'ISS', 'ISS', Decimal('13.760'), Decimal('-113.210'), 'ISS'),
 (datetime.date(2012, 7, 15), 'Akihiko Hoshide', 'ISS', 'ISS', Decimal('126.970'), Decimal('0.000'), 'ISS'),
 (datetime.date(1969, 11, 14), 'Alan Bean', 'Moon', None, Decimal('10.190'), Decimal('-49.280'), 'NV'),
 (datetime.date(1973, 7, 28), 'Alan Bean', 'Other Space station', 'Skylab', Decimal('59.470'), Decimal('0.000'), 'Skylab'),
 (datetime.date(2008, 2, 7), 'Alan G. Poindexter', 'ISS', 'ISS', Decimal('12.770'), Decimal('-2.350'), 'ISS'),
 (datetime.date(2010, 4, 5), 'Alan G. Poindexter', 'ISS', 'ISS', Decimal('15.120'), Decimal('0.000'), 'ISS'),
 (datetime.date(1961, 5, 5), 'Alan Shepard', 'Non-orbital space', None, Decimal('0.010'), Decimal('-8.990'), 'NV'),
 (datetime.date(1971, 1, 31), 'Alan Shepard', 'Moon', None, Decimal('9.000'), Decimal('0.000'), 'NV'),
 (datetime.date(1995, 10, 20), 'Albert Sacco', 'Earth Orbit', None, Decimal('15.910'), Decimal('0.000'), 'NV'),
 (datetime.date(1990, 2, 11), 'Aleksandr Balandin', 'Mir', 'Mir', Decimal('179.050'), Decimal('0.000'), 'Mir'),
 (datetime.date(1978, 6, 15), 'Aleksandr Ivanchenkov', 'Other Space station', 'Salyut 6', Decimal('139.620'), Decimal('131.710'), 'Salyut 6'),
 (datetime.date(1982, 6, 24), 'Aleksandr Ivanchenkov', 'Other Space station', 'Salyut 7', Decimal('7.910'), Decimal('0.000'), 'Salyut 7'),
 (datetime.date(1987, 2, 5), 'Aleksandr Laveykin', 'Mir', 'Mir', Decimal('174.140'), Decimal('0.000'), 'Mir'),
 (datetime.date(1997, 2, 10), 'Aleksandr Lazutkin', 'Mir', 'Mir', Decimal('184.920'), Decimal('0.000'), 'Mir'),
 (datetime.date(1988, 6, 7), 'Aleksandr Panayotov Aleksandrov', 'Mir', 'Mir', Decimal('9.520'), Decimal('0.000'), 'Mir'),
 (datetime.date(1983, 6, 27), 'Aleksandr Pavlovich Aleksandrov', 'Other Space station', 'Salyut 7', Decimal('149.450'), Decimal('-10.850'), 'Salyut 7'),
 (datetime.date(1987, 7, 22), 'Aleksandr Pavlovich Aleksandrov', 'Mir', 'Mir', Decimal('160.300'), Decimal('0.000'), 'Mir'),
 (datetime.date(1993, 1, 24), 'Aleksandr Poleshchuk', 'Mir', 'Mir', Decimal('179.030'), Decimal('0.000'), 'Mir'),
 (datetime.date(2011, 4, 4), 'Aleksandr Samokutyayev', 'ISS', 'ISS', Decimal('164.240'), Decimal('-3.000'), 'ISS'),
 (datetime.date(2014, 9, 25), 'Aleksandr Samokutyayev', 'ISS', 'ISS', Decimal('167.240'), Decimal('0.000'), 'ISS'),
 (datetime.date(1982, 8, 19), 'Aleksandr Serebrov', 'Other Space station', 'Salyut 7', Decimal('7.910'), Decimal('5.900'), 'Salyut 7'),
 (datetime.date(1983, 4, 20), 'Aleksandr Serebrov', 'Earth Orbit', None, Decimal('2.010'), Decimal('-164.280'), 'NV'),
 (datetime.date(1989, 9, 5), 'Aleksandr Serebrov', 'Mir', 'Mir', Decimal('166.290'), Decimal('-30.450'), 'Mir'),
 (datetime.date(1993, 7, 1), 'Aleksandr Serebrov', 'Mir', 'Mir', Decimal('196.740'), Decimal('0.000'), 'Mir'),
 (datetime.date(2010, 4, 2), 'Aleksandr Skvortsov', 'ISS', 'ISS', Decimal('176.050'), Decimal('6.840'), 'ISS'),
 (datetime.date(2014, 3, 25), 'Aleksandr Skvortsov', 'ISS', 'ISS', Decimal('169.210'), Decimal('-31.490'), 'ISS'),
 (datetime.date(2019, 7, 20), 'Aleksandr Skvortsov', 'ISS', 'ISS', Decimal('200.700'), Decimal('0.000'), 'ISS'),
 (datetime.date(1987, 7, 22), 'Aleksandr Viktorenko', 'Mir', 'Mir', Decimal('7.960'), Decimal('-158.330'), 'Mir'),
 (datetime.date(1989, 9, 5), 'Aleksandr Viktorenko', 'Mir', 'Mir', Decimal('166.290'), Decimal('20.700'), 'Mir'),
 (datetime.date(1992, 3, 17), 'Aleksandr Viktorenko', 'Mir', 'Mir', Decimal('145.590'), Decimal('-23.630'), 'Mir'),
 (datetime.date(1994, 10, 3), 'Aleksandr Viktorenko', 'Mir', 'Mir', Decimal('169.220'), Decimal('0.000'), 'Mir'),
 (datetime.date(1985, 9, 17), 'Aleksandr Volkov', 'Other Space station', 'Salyut 7', Decimal('64.910'), Decimal('-86.550'), 'Salyut 7'),
 (dat

In [48]:
%%sql
SELECT CrewMissionName,Name,StartDate,EndDate,Birthday,DEEP
FROM
(SELECT ST.* ,SW.CrewMissionName,SW.StartDate,SW.EndDate, CASE WHEN SW.DeepSpace = TRUE THEN "OK"
END DEEP
FROM SpaceWalks SW
LEFT JOIN SpaceTravelers ST
ON ST.Name = SW.Name
WHERE  EVA_Number >= 3 AND EVA_Number <= 15) AS S
GROUP BY Name
LIMIT 10;



 * mysql://root:***@localhost
10 rows affected.


[('Gemini 9', 'Gene Cernan', datetime.date(1966, 6, 5), datetime.date(1966, 6, 5), datetime.date(1934, 3, 14), None),
 ('Gemini 10', 'Michael Collins', datetime.date(1966, 7, 19), datetime.date(1966, 7, 19), datetime.date(1930, 10, 31), None),
 ('Gemini 11', 'Richard Gordon', datetime.date(1966, 9, 13), datetime.date(1966, 9, 13), datetime.date(1929, 10, 5), None),
 ('Gemini 12', 'Buzz Aldrin', datetime.date(1966, 11, 12), datetime.date(1966, 11, 12), datetime.date(1930, 1, 20), None),
 ('Soyuz 4/5', 'Yevgeny Khrunov', datetime.date(1969, 1, 16), datetime.date(1969, 1, 16), datetime.date(1933, 9, 10), None),
 ('Soyuz 4/5', 'Aleksei Yeliseyev', datetime.date(1969, 1, 16), datetime.date(1969, 1, 16), datetime.date(1934, 7, 13), None),
 ('Apollo 9', 'Rusty Schweickart', datetime.date(1969, 3, 6), datetime.date(1969, 3, 6), datetime.date(1935, 10, 25), None),
 ('Apollo 9', 'David Scott', datetime.date(1969, 3, 6), datetime.date(1969, 3, 6), datetime.date(1932, 6, 6), None),
 ('Apollo 11', 'Neil Armstrong', datetime.date(1969, 7, 21), datetime.date(1969, 7, 21), datetime.date(1930, 8, 5), None),
 ('Apollo 12', 'Pete Conrad', datetime.date(1969, 11, 19), datetime.date(1969, 11, 19), datetime.date(1930, 5, 2), None)]

##### This gives the details of the Crewmission name, The name of the crew members, their take off date and return date along with their date of Birth with the conditions where `DeepSpace` is True and `Eva Number is minimum of 3 and maximum of 15`

In [50]:
%%sql
/***Testing purpose******/
SELECT * , CASE WHEN DeepSpace = FALSE THEN "OK"
END 
FROM SpaceWalks
GROUP BY 1
LIMIT 50;

 * mysql://root:***@localhost
50 rows affected.


[('Voskhod 2', 'Alexei Leonov', 1, datetime.date(1965, 3, 18), datetime.date(1965, 3, 18), datetime.timedelta(seconds=729), 0, 0, 0, datetime.date(1965, 3, 18), 'Voskhod 2', 'Earth Orbit', 'OK'),
 ('Gemini 4', 'Ed White', 2, datetime.date(1965, 6, 3), datetime.date(1965, 6, 3), datetime.timedelta(seconds=1200), 0, 0, 0, datetime.date(1965, 6, 3), 'Gemini 4', 'Earth Orbit', 'OK'),
 ('Gemini 9', 'Gene Cernan', 3, datetime.date(1966, 6, 5), datetime.date(1966, 6, 5), datetime.timedelta(seconds=7620), 0, 0, 0, datetime.date(1966, 6, 3), 'Gemini 9', 'Earth Orbit', 'OK'),
 ('Gemini 10', 'Michael Collins', 4, datetime.date(1966, 7, 19), datetime.date(1966, 7, 19), datetime.timedelta(seconds=2940), 0, 0, 0, datetime.date(1966, 7, 18), 'Gemini 10', 'Earth Orbit', 'OK'),
 ('Gemini 11', 'Richard Gordon', 6, datetime.date(1966, 9, 13), datetime.date(1966, 9, 13), datetime.timedelta(seconds=1980), 0, 0, 0, datetime.date(1966, 9, 12), 'Gemini 11', 'Earth Orbit', 'OK'),
 ('Gemini 12', 'Buzz Aldrin', 8, datetime.date(1966, 11, 12), datetime.date(1966, 11, 12), datetime.timedelta(seconds=8940), 0, 0, 0, datetime.date(1966, 11, 11), 'Gemini 12', 'Earth Orbit', 'OK'),
 ('Soyuz 4/5', 'Yevgeny Khrunov', 11, datetime.date(1969, 1, 16), datetime.date(1969, 1, 16), datetime.timedelta(seconds=1920), 0, 0, 0, datetime.date(1969, 1, 15), 'Soyuz 5', 'Earth Orbit', 'OK'),
 ('Apollo 9', 'Rusty Schweickart', 12, datetime.date(1969, 3, 6), datetime.date(1969, 3, 6), datetime.timedelta(seconds=4620), 0, 0, 0, datetime.date(1969, 3, 3), 'Apollo 9', 'Earth Orbit', 'OK'),
 ('Apollo 11', 'Neil Armstrong', 13, datetime.date(1969, 7, 21), datetime.date(1969, 7, 21), datetime.timedelta(seconds=9100), 1, 0, 0, datetime.date(1969, 7, 16), 'Apollo 11', 'Moon', 'OK'),
 ('Apollo 12', 'Pete Conrad', 14, datetime.date(1969, 11, 19), datetime.date(1969, 11, 19), datetime.timedelta(seconds=14163), 1, 0, 0, datetime.date(1969, 11, 14), 'Apollo 12', 'Moon', 'OK'),
 ('Apollo 14', 'Alan Shepard', 16, datetime.date(1971, 2, 5), datetime.date(1971, 2, 5), datetime.timedelta(seconds=17270), 1, 0, 0, datetime.date(1971, 1, 31), 'Apollo 14', 'Moon', 'OK'),
 ('Apollo 15', 'David Scott', 18, datetime.date(1971, 7, 31), datetime.date(1971, 7, 31), datetime.timedelta(seconds=1987), 0, 1, 0, datetime.date(1971, 7, 26), 'Apollo 15', 'Moon', 'OK'),
 ('Apollo 16', 'John Young', 23, datetime.date(1972, 4, 21), datetime.date(1972, 4, 21), datetime.timedelta(seconds=25872), 1, 0, 0, datetime.date(1972, 4, 16), 'Apollo 16', 'Moon', 'OK'),
 ('Apollo 17', 'Gene Cernan', 27, datetime.date(1972, 12, 11), datetime.date(1972, 12, 12), datetime.timedelta(seconds=25913), 1, 0, 0, datetime.date(1972, 12, 7), 'Apollo 17', 'Moon', 'OK'),
 ('Skylab 2\xa0CM', 'Paul Weitz', 31, datetime.date(1973, 5, 26), datetime.date(1973, 5, 26), datetime.timedelta(seconds=2400), 0, 0, 0, datetime.date(1973, 5, 25), 'Skylab 2', 'Other Space station', 'OK'),
 ('Skylab 2 - EVA 2', 'Pete Conrad', 32, datetime.date(1973, 6, 7), datetime.date(1973, 6, 7), datetime.timedelta(seconds=12300), 0, 0, 0, datetime.date(1973, 5, 25), 'Skylab 2', 'Other Space station', 'OK'),
 ('Skylab 2 - EVA 3', 'Pete Conrad', 33, datetime.date(1973, 6, 19), datetime.date(1973, 6, 19), datetime.timedelta(seconds=5760), 0, 0, 0, datetime.date(1973, 5, 25), 'Skylab 2', 'Other Space station', 'OK'),
 ('Skylab 3', 'Owen Garriott', 34, datetime.date(1973, 8, 6), datetime.date(1973, 8, 7), datetime.timedelta(seconds=23460), 0, 0, 0, datetime.date(1973, 7, 28), 'Skylab 3', 'Other Space station', 'OK'),
 ('Skylab 4', 'Edward Gibson', 37, datetime.date(1973, 11, 22), datetime.date(1973, 11, 23), datetime.timedelta(seconds=23580), 0, 0, 0, datetime.date(1973, 11, 16), 'Skylab 4', 'Other Space station', 'OK'),
 ('Salyut 6 – PE-1', 'Yuri Romanenko', 41, datetime.date(1977, 12, 19), datetime.date(1977, 12, 19), datetime.timedelta(seconds=5280), 0, 1, 0, datetime.date(1977, 12, 10), 'Soyuz 26', 'Other Space station', 'OK'),
 ('Salyut 6 – PE-2', 'Vladimir Kovalyonok

In [49]:
%%sql
/*****Testing purpose*******/
SELECT Name,Moonwalk, EndDate
FROM 
Spacewalks
WHERE EVA_Number >= 3 AND EVA_Number <= 15

 * mysql://root:***@localhost
18 rows affected.


[('Gene Cernan', 0, datetime.date(1966, 6, 5)),
 ('Michael Collins', 0, datetime.date(1966, 7, 19)),
 ('Michael Collins', 0, datetime.date(1966, 7, 20)),
 ('Richard Gordon', 0, datetime.date(1966, 9, 13)),
 ('Richard Gordon', 0, datetime.date(1966, 9, 14)),
 ('Buzz Aldrin', 0, datetime.date(1966, 11, 12)),
 ('Buzz Aldrin', 0, datetime.date(1966, 11, 13)),
 ('Buzz Aldrin', 0, datetime.date(1966, 11, 14)),
 ('Yevgeny Khrunov', 0, datetime.date(1969, 1, 16)),
 ('Aleksei Yeliseyev', 0, datetime.date(1969, 1, 16)),
 ('Rusty Schweickart', 0, datetime.date(1969, 3, 6)),
 ('David Scott', 0, datetime.date(1969, 3, 6)),
 ('Neil Armstrong', 1, datetime.date(1969, 7, 21)),
 ('Buzz Aldrin', 1, datetime.date(1969, 7, 21)),
 ('Pete Conrad', 1, datetime.date(1969, 11, 19)),
 ('Alan Bean', 1, datetime.date(1969, 11, 19)),
 ('Pete Conrad', 1, datetime.date(1969, 11, 20)),
 ('Alan Bean', 1, datetime.date(1969, 11, 20))]